FGEGraphics/GraphicsHelpers/Models/ShapeGenerators.cs

```diff
@@ -0,0 +1,408 @@
+using OpenTK.Mathematics;
+using FGECore.CoreSystems;
+
+namespace FGEGraphics.GraphicsHelpers.Models;
+
+/// <summary>Generates shapes and returns them as a model.</summary>
+public static class ShapeGenerators
+{
+    /// <summary>
+    /// Generates a 3D sphere model.
+    /// Note that the stacks should be lower than the slices to avoid incorrect normals a good example is 10 stacks and 60 slices.
+    /// </summary>
+    /// <param name="radius"></param>
+    /// <param name="stacks"></param>
+    /// <param name="slices"></param>
+    /// <param name="modelEngine"></param>
+    /// <param name="reverseOrder"></param>
+    /// <returns>The generated model.</returns>
+    public static Model GenerateSphere(float radius, uint stacks, uint slices, ModelEngine modelEngine, bool reverseOrder = false)
+    {
+        // Calculate the number of vertices and indices
+        uint vertexCount = (stacks + 1) * (slices + 1);
+        uint numIndices = stacks * slices * 6;  // 2 triangles per stack/slice, 3 indices per triangle
+
+        List<Vector3> vertices = new((int)vertexCount);
+        List<Vector3> normals = new((int)vertexCount);
+        List<Vector2> texCoords = new((int)vertexCount);
+
+        uint[] indices = new uint[numIndices];
+        int index = 0;
+
+        for (int i = 0; i <= stacks; i++)
+        {
+            float theta = i * MathHelper.Pi / stacks;
+            float sinTheta = (float)Math.Sin(theta);
+            float cosTheta = (float)Math.Cos(theta);
+
+            for (int j = 0; j <= slices; j++)
+            {
+                float phi = j * 2 * MathHelper.Pi / slices;
+                float sinPhi = (float)Math.Sin(phi);
+                float cosPhi = (float)Math.Cos(phi);
+
+                float x = radius * cosPhi * sinTheta;
+                float y = radius * cosTheta;
+                float z = radius * sinPhi * sinTheta;
+
+                vertices.Add(new Vector3(x, y, z));
+
+                Vector3 normal = new(x, y, z);
+                normals.Add(normal.Normalized());
+
+                texCoords.Add(new Vector2((float)j / slices, (float)i / stacks));
+
+                if (i < stacks && j < slices)
+                {
+                    uint currentRow = (uint)(i * (slices + 1));
+                    uint nextRow = (uint)((i + 1) * (slices + 1));
+
+                    indices[index++] = (uint)(currentRow + j);
+                    indices[index++] = (uint)(nextRow + j);
+                    indices[index++] = (uint)(currentRow + j + 1);
+
+                    indices[index++] = (uint)(currentRow + j + 1);
+                    indices[index++] = (uint)(nextRow + j);
+                    indices[index++] = (uint)(nextRow + j + 1);
+                }
+            }
+        }
+
+        if (vertices.Count != normals.Count || vertices.Count != texCoords.Count)
+            throw new ArgumentException("Position, normal, and texture coordinate lists must have the same length.");
```

Comment:
formatting, should have `{}` on all blocks

Reply:
Also what is this error checking? This seems more like a debug assert over your own code rather than an argument check for the caller